openmmmcmc/states.py

```diff
@@ -0,0 +1,1903 @@
+    particles of a subset of atoms
+
+    >>> sliced_sampler_state = sampler_state[:10]
+    >>> sliced_sampler_state.n_particles
+    10
+
+    """
+
+    # -------------------------------------------------------------------------
+    # Public interface
+    # -------------------------------------------------------------------------
+
+    def __init__(self, positions, velocities=None, box_vectors=None):
+        self._positions = positions
+        self._cached_positions_in_md_units = None
+        self._velocities = None
+        self._cached_velocities_in_md_units = None
+        self.velocities = velocities  # Checks consistency and units.
+        self.box_vectors = box_vectors
+        self.potential_energy = None
+        self.kinetic_energy = None
+
+    @staticmethod
+    def from_context(context):
+        """Alternative constructor.
+
+        Read all the configurational properties from a Context object.
+        This guarantees that all attributes (including energy attributes)
+        are initialized.
+
+        Parameters
+        ----------
+        context : simtk.openmm.Context
+            The context to read.
+
+        Returns
+        -------
+        sampler_state : SamplerState
+            A new SamplerState object.
+
+        """
+        sampler_state = SamplerState([])
+        sampler_state._read_context_state(context, check_consistency=False)
+        return sampler_state
+
+    @property
+    def positions(self):
+        """Particle positions.
+
+        An Nx3 simtk.unit.Quantity object, where N is the number of
+        particles.
+
+        Raises
+        ------
+        SamplerStateError
+            If set to an array with a number of particles different
+            than n_particles.
+
+        """
+        return self._positions
+
+    @positions.setter
+    def positions(self, value):
+        if value is None or len(value) != self.n_particles:
+            raise SamplerStateError(SamplerStateError.INCONSISTENT_POSITIONS)
+        self._positions = value
+        self._cached_positions_in_md_units = None  # Invalidate cache.
+
+    @property
+    def velocities(self):
+        """Particle velocities.
+
+        An Nx3 simtk.unit.Quantity object, where N is the number of
+        particles.
+
+        Raises
+        ------
+        SamplerStateError
+            If set to an array with a number of particles different
+            than n_particles.
+
+        """
+        return self._velocities
+
+    @velocities.setter
+    def velocities(self, value):
+        if value is not None and self.n_particles != len(value):
+            raise SamplerStateError(SamplerStateError.INCONSISTENT_VELOCITIES)
+        self._velocities = value
+        self._cached_velocities_in_md_units = None  # Invalidate cache.
+
+    @property
+    def total_energy(self):
+        """The sum of potential and kinetic energy (read-only)."""
+        if self.potential_energy is None or self.kinetic_energy is None:
```

Comment:
I thought about returning `None` if only one of them is `None` to avoid having ambiguous results (e.g. when the kinetic energy is actually 0 for example).